File: libsql/src/main/java/tech/turso/libsql/Libsql.java
package tech.turso.libsql;

public class Libsql {

    // Used to load the 'libsql' library on application startup.
    static {
        System.loadLibrary("libsql_android");
    }

    public static Database openLocal(String dbFile) throws Exception {
        return new Database(nativeOpenLocal(dbFile));
    }

    public static Database openRemote(String url, String authToken) throws Exception {
        return new Database(nativeOpenRemote(url, authToken));
    }

    public static EmbeddedReplicaDatabase openEmbeddedReplica(
            String dbFile, String url, String authToken) throws Exception {
        return new EmbeddedReplicaDatabase(nativeOpenEmbeddedReplica(dbFile, url, authToken));
    }

    private static native long nativeOpenLocal(String dbFile) throws Exception;

    private static native long nativeOpenRemote(String url, String authToken);

    private static native long nativeOpenEmbeddedReplica(
            String dbFile, String url, String authToken);
}

File: libsql/src/main/kotlin/tech/turso/libsql/Libsql.kt
package tech.turso.libsql

import com.google.protobuf.ByteString
import tech.turso.libsql.proto.Value

fun Any?.toValue(): Value = when (this) {
    is Int -> Value.newBuilder().setInteger(this.toLong()).build()
    is Long -> Value.newBuilder().setInteger(this).build()
    is String -> Value.newBuilder().setText(this).build()
    is Float -> Value.newBuilder().setReal(this.toDouble()).build()
    is Double -> Value.newBuilder().setReal(this).build()
    is ByteArray -> Value.newBuilder().setBlob(ByteString.copyFrom(this)).build()
    null -> Value.newBuilder().setNull(Value.Null.newBuilder().build()).build()
    else -> {
        throw IllegalArgumentException("${this::class.simpleName} not supported")
    }
}

object Libsql {
    init {
        System.loadLibrary("libsql_android")
    }

    @JvmStatic
    fun open(path: String): Database {
        return Database(nativeOpenLocal(path))
    }

    @JvmStatic
    fun open(
        url: String,
        authToken: String,
    ): Database {
        return Database(nativeOpenRemote(url, authToken))
    }

    @JvmStatic
    fun open(
        path: String,
        url: String,
        authToken: String,
        syncInterval: Long = 0,
        readYourWrites: Boolean = true,
        withWebpki: Boolean = true,
    ): EmbeddedReplicaDatabase {
        return EmbeddedReplicaDatabase(
            nativeOpenEmbeddedReplica(
                path,
                url,
                authToken,
                syncInterval,
                readYourWrites,
                withWebpki
            )
        )
    }

    private external fun nativeOpenLocal(path: String): Long

    private external fun nativeOpenRemote(
        url: String,
        authToken: String,
    ): Long

    private external fun nativeOpenEmbeddedReplica(
        path: String,
        url: String,
        authToken: String,
        syncInterval: Long,
        readYourWrites: Boolean,
        withWebpki: Boolean,
    ): Long
}
